platform/wayland/cog-platform-wl.c

```diff
@@ -1560,16 +1579,42 @@ static struct wpe_view_backend *
 gamepad_provider_get_view_backend_for_gamepad(void *provider G_GNUC_UNUSED, void *gamepad G_GNUC_UNUSED)
 {
     CogWlPlatform *platform = COG_WL_PLATFORM(cog_platform_get_default());
-    CogWlView     *view = platform->view;
+
+    CogWlView *view = (CogWlView *) cog_viewport_get_visible_view(platform->viewport);
     g_assert(view);
     return wpe_view_backend_exportable_fdo_get_view_backend(view->exportable);
 }
 
+static void
+cog_wl_platform_on_notify_visible_view(CogWlPlatform *self)
+{
+    CogWlView *view = (CogWlView *) cog_viewport_get_visible_view(self->viewport);
+    g_debug("%s: Visible view %p.", G_STRFUNC, view);
+
+    if (!(view && view->image)) {
+        g_debug("%s: No image to show, skipping update.", G_STRFUNC);
+        return;
+    }
+
+    /*
+     * TODO: Once we support multiple viewports, a view may be visible
+     *       without having focus. At that point the input events will
+     *       need to go to the *focused* view instead.
+     *
+     *       For now add the flag and assume that visible == focused.
+     */
+    wpe_view_backend_add_activity_state(cog_view_get_backend((CogView *) view), wpe_view_activity_state_focused);
+    cog_wl_view_update_surface_contents(view);
```

Comment:
Related to https://github.com/Igalia/cog/pull/644#issuecomment-1792067605. You want to set the focus in the visible view always once this is visible (this could change in the future with multiple viewports and so but now it is the way to set the focus in the visible View). The check for view->image only matters for refresh the content with a new frame.

 
```suggestion
    /*
     * TODO: Once we support multiple viewports, a view may be visible
     *       without having focus. At that point the input events will
     *       need to go to the *focused* view instead.
     *
     *       For now add the flag and assume that visible == focused.
     */
    wpe_view_backend_add_activity_state(cog_view_get_backend((CogView *) view), wpe_view_activity_state_focused);

    if (!(view && view->image)) {
        g_debug("%s: No image to show, skipping update.", G_STRFUNC);
        return;
    }

    cog_wl_view_update_surface_contents(view);
```

Reply:
Oh, wow, excellent catch. Indeed we always want to set the flag, and not only when there is an image to update the surface 🖼️ 

I will update the PR with this change, make the style checker happy, and re-upload the patch.

Thanks!

Reply:
I think this is ok for now since we will want the focus always activated  in the visible view but in the future it will be necessary to differentiate between these to different concepts (visibility vs focus).

Look at this: `cog set visible view 3 from viewport 2` VS `cog set focus on viewport 2`. The first case should activate the focus on view 3 of the viewport 2 only if the focus was previously active on viewport 2. But it will be not always the case. For example, we may want to make view N of a viewport visible but not steal the focus  from another viewport which is the one with focus at this point.